### PR TITLE
Split BuffWindow into smaller components

### PR DESCRIPTION
**Pull request by KelosPelmand**
Existing BuffWindows have been converted to the new system.  This system now enables
- Windows based on things other than buffs (eg Hypercharge).
- Creation of custom evaulations for BuffWindow (and other windows) without having to modify the core class.
- Tracking of a history of windows via a common class that contains the boilerplate (History).
- (Also, is now an Analyser.)

Fun fact: TrickAttack does not need a special hook in the new system since statusApply/Remove in Analyser space includes both buffs and debuffs where the similar events in Module space were buff only.

I tested a random sample of parses for each class, but if you have parses that are known to contain edge cases that were the reason we added overrides to old BuffWindow in the first place, please feel free to check those against this new system to ensure I did not miss anything.

### CODE REVIEW

**Comment by ayyaruq** [comment]
Does this address #1158 at all? This could be useful in 6.0 for MNK, however our window is status stack based (within a time constraint), and needs to show only one of a potential list in column data, possibly with an image of the skill or even gauge element, rather than numeric usages.

**Comment by KelosPelmand** [comment]
Yes and no.
No, it does not directly implement a fix to #1158.
Yes, it does allow you to implement a fix by making a new evaluator class.  Initial thought for a name would be ExpectedActionGroupEvaluator and it would be similar to ExpectedActionsEvalutor except that it outputs a single column in the table for all actions together, rather than one column per action.

File: src/parser/jobs/pld/modules/Gauge.tsx [hunk 1]
```diff
@@ -55,7 +55,6 @@ export default class Gauge extends CoreGauge {
 				A total of {this.oathGauge.overCap} gauge was lost due to exceeding the cap.
 			</Trans>,
 			severity: SEVERITY.MINOR,
-			value: this.oathGauge.overCap,
```
Comment (AzariahAeroborn) [inline src/parser/jobs/pld/modules/Gauge.tsx:58]: This shouldn't be removed, and not part of a Buff Window change either.

File: src/parser/jobs/pld/modules/FightOrFlight.tsx [hunk 2]
```diff
@@ -229,7 +228,6 @@ export default class FightOrFlight extends Module {
 
 		this.suggestions.add(new TieredSuggestion({
 			icon: ACTIONS.GORING_BLADE.icon,
-			severity: matchClosestLower(SEVERITIES.MISSED_GCD, this.fofErrorResult.goringTooCloseCounter),
```
Comment (AzariahAeroborn) [inline src/parser/jobs/pld/modules/FightOrFlight.tsx:232]: This seems like an unrelated change that shouldn't be in this pull request. (And kind of seems like an unintended change that shouldn't be being removed).

**Review by AzariahAeroborn — CHANGES_REQUESTED** [review]
Noticed these while skimming, the gauge one stuck out to me because of a WAR gauge bug report I was looking at.

File: src/parser/jobs/pld/modules/FightOrFlight.tsx (hunk 2 above)
Comment (KelosPelmand) [reply src/parser/jobs/pld/modules/FightOrFlight.tsx:232]: This value was already being ignored because it is a TieredSuggestion, unless I misunderstand how JS was working.  (I didn't notice any difference in values for the PLD parses I checked.)  TieredSuggestion calculates its severity (using this matchClosestLower function interally) based on the tiers and value properties.  This had to be removed since Suggestion file was converted to TS and now takes a constructor object instead of an any.

File: src/parser/jobs/pld/modules/Gauge.tsx (hunk 1 above)
Comment (KelosPelmand) [reply src/parser/jobs/pld/modules/Gauge.tsx:58]: As above, this value had to be removed since Suggestion file was converted to TS and now takes a constructor object instead of an any.  The value property is part of TieredSuggestion but is not used by Suggestion.

File: src/parser/jobs/pld/modules/FightOrFlight.tsx (hunk 2 above)
Comment (AzariahAeroborn) [reply src/parser/jobs/pld/modules/FightOrFlight.tsx:232]: Ah, whoops, I misread the surrounding context of the change.

File: src/parser/jobs/drg/modules/DragonSight.tsx [hunk 3]
```diff
@@ -14,117 +19,158 @@ import DISPLAY_ORDER from './DISPLAY_ORDER'
...
-				action: ACTIONS.CHAOS_THRUST,
-				expectedPerWindow: 1,
-			},
-			{
-				action: ACTIONS.FULL_THRUST,
-				expectedPerWindow: 1,
+	override initialise() {
+		super.initialise()
+
+		this.addEventHook({type: 'death', to: 'player'}, this.onDeath)
```
Comment (ackwell) [inline src/parser/jobs/drg/modules/DragonSight.tsx:87]: breaking; this won't fire on the new event system

**Review by ackwell — CHANGES_REQUESTED** [review]
big pr gets big comments
there's a lot there, sorry, but i tried to keep the average under 1 per file
i've tried to make it clear what's requested and what's a nice to have, reach out if questions and stuff

i was going to take a look at fixing some myself but running low on time on it for today, it took longer to review than i expected.

File: src/parser/core/modules/History.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	data: T
+}
+
+export class History<T> {
+
+	/**
+	 * Creates a tracker with the specified initialization function
+	 * @param initializer Creates data for a new entry when one is opened.
+	 */
+	constructor(initializer: () => T) {
```
[the remark this answers is not in the document]
Comment (ayyaruq) [reply src/parser/core/modules/History.ts:24]: Does the linter allow the z here? Should it? Thinking consistency mostly.

File: src/parser/core/modules/Suggestions/Suggestion.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	content: ReactNode
+	why: ReactNode
+	severity?: number
+}
+
+export default class Suggestion {
+	public icon: string // TODO: default image
+	public content: ReactNode
+	public why: ReactNode
+	private severityValue?: number
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/Suggestions/Suggestion.ts:25]: The undefined option is basically there because of how TieredSuggestion passes things into the constructor.  However, it doesn't really need to be there for the base class field otherwise, since non-tiered Suggestions should always be passing one in.  Should be able to tweak these around a bit to get rid of it here.

File: src/parser/core/modules/ActionWindow/readme.md [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+	/**
+	 * Determines if a window ended early due to the end of the pull.
+	 * @param window The window to check
+	 * @returns True if the window is shorter than the expected duration of the buff because of the end
+	 * of the pull; false otherwise.
+	 */
+	protected isRushedEndOfPullWindow(window: HistoryEntry<EvaluatedAction[]>) { /* .. */ }
+
+    // Very important, you must call super.initialise() in any implementing module.
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/readme.md:26]: Looks like I edited this in something configured to use spaces instead of tabs; will be fixed.

File: src/parser/core/modules/ActionWindow/readme.md [hunk 7]
```diff
@@ -0,0 +1,276 @@
+# BuffWindow Interface
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/readme.md:1]: The original version of this writeup was specifically talking about an upgrade of BuffWindow and had much more comparison to the existing version, so starting with BuffWindow seemed natural.  It's also the implementation that most people will be interacting with, so I left it as the top.  Let me know if you would like to see a different order.

> Also please use single space after full stop. It's 2021 😔

Yes it is in fact 2021.  I did not see any messages indicating that the use of the double space would expire before 2021 (or ever in fact), so I and many people I know still use the double space after full stop as we were once taught.

But, style rules for a project are the style rules, so it shall be changed.

File: src/parser/core/modules/History.ts [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	public entries: Array<HistoryEntry<T>> = []
+
+	private dataInitializer: () => T
+
+	/**
+	 * Determines if there is a currently open entry.
+	 */
+	public isEntryOpen() {
+		const last = _.last(this.entries)
+		return last != null && last.end == null
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/History.ts:40]: Looking at it more, isEntryOpen is leftover from my first pass at the class and doesn't really serve any purpose anymore.  There's no reason I can think of to want to know if an entry is open and not then immediately interact with said open entry.

File: src/parser/core/modules/History.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+	/**
+	 * Closes the current entry.
+	 * If the most recent entry is already closed, no action is taken
+	 * @param timestamp The timestamp at which the entry is closed.
+	 */
+	public closeCurrent(timestamp: number) {
+		const last = _.last(this.entries)
+		if (last == null || last.end != null) { return }
+		last.end = timestamp
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/History.ts:95]: I had to look at this twice to catch what it was doing.  `current && current.end = timestamp` reads as something that should return a value at first glance.  How would you feel about:

`if (current != null) { current.end = timestamp } `

instead?

File: src/parser/core/modules/History.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	 * The timestamp at which the entry ended or undefined if the entry is open.
+	 */
+	end?: number
+	/**
+	 * The data for this entry.
+	 */
+	data: T
+}
+
+export class History<T> {
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/History.ts:18]: I don't think this is best placed in the ActionWindow folder.  It is intended to be useful outside of ActionWindow, and in fact, is already being discussed to be used by RDM for it's melee combo analyser.  Would `src/parser/core make` sense instead?

File: src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+import {WindowEvaluator} from '../evaluators/WindowEvaluator'
+
+/**
+ * Tracks actions that occur within a window.
+ * By default, all actions cast during a window will be included.
+ */
+export abstract class ActionWindow extends Analyser {
+
+	@dependency protected data!: Data
+	@dependency suggestions!: Suggestions
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx:23]: All suggestions come from the evaluators, and the derived window classes don't output any suggestions of their own, so I don't see any reason to make it protected.

File: src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+	}
+
+	private onComplete() {
+		this.onWindowEnd(this.parser.pull.timestamp + this.parser.pull.duration)
+
+		const actionHistory = this.mapHistoryActions()
+		this.evaluators
+			.map(ev => ev.suggest(actionHistory))
+			.filter(isDefined)
+			.forEach(s => this.suggestions.add(s))
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx:142]: Easy enough.  For the future, is there a baseline to use to decide when to use a loop array.x calls?  Is it a matter of how many you are chaining together?

File: src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+			notes={notesData}
+			onGoto={this.timeline.show}
+			headerTitle={this.rotationTableHeader}
+		/>
+	}
+
+	private mapHistoryActions(): Array<HistoryEntry<EvaluatedAction[]>> {
+		return this.history.entries
+			.map(entry => {
+				return {start: entry.start,
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx:191]: Ah, the parens are the magic I was missing.  I was trying to just put the literal there and it wouldn't work without the return.

File: src/parser/core/modules/ActionWindow/readme.md [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+
+    // Very important, you must call super.initialise() in any implementing module.
+	override initialise() { /* .. */ }
+
+}
+```
+
+One more override property has snuck in here, the override for the header of the RotationTable in the output.
+
+The most important function here is `addEvaluator`.  There are a variety of pre-built WindowEvaluators that cover the common evaluations.  Implementing modules will need to call `addEvaulator` for each evaluation they want to perform in their `initialise` method.
```
Comment (hintxiv) [inline src/parser/core/modules/ActionWindow/readme.md:109]:
```suggestion
The most important function here is `addEvaluator`.  There are a variety of pre-built WindowEvaluators that cover the common evaluations.  Implementing modules will need to call `addEvaluator` for each evaluation they want to perform in their `initialise` method.
```

File: src/parser/core/modules/ActionWindow/windows/BuffWindow.ts [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+	override initialise() {
+		super.initialise()
+
+		// need to check source and target due to buff mirroring. Check for noneOf pets
+		// to allow debuff based windows such as Trick Attack to work through this class.
+		const pets = this.parser.pull.actors
+			.filter(actor => actor.owner === this.parser.actor)
+			.map(actor => actor.id)
+		const playerFilter = filter<Event>().source(this.parser.actor.id).target(noneOf(pets))
+		const buffFilter = playerFilter.status(oneOf(ensureArray(this.buffStatus).map(s => s.id)))
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/windows/BuffWindow.ts:50]: That comment isn't even right.  The pet part of the filter is for buff mirroring, not debuff mirroring.  I probably had some other comment there before (most likely about how Trick Attack works without special hooks now) and didn't pay attention when I wrote about the pets.

File: src/parser/core/modules/ActionWindow/evaluators/LimitedActionsEvaluator.tsx [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+				return total + missingInWindow
+			}, 0)
+
+		return new TieredSuggestion({
+			icon: this.suggestionIcon,
+			content: this.suggestionContent,
+			tiers: this.severityTiers,
+			value: extraCount,
+			why: <Trans id="core.buffwindow.suggestions.trackedbadaction.why">
+				<Plural value={extraCount} one="# use of" other="# uses of"/> actions that should be avoided during {this.windowName} windows.
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/evaluators/LimitedActionsEvaluator.tsx:47]: Window name is currently always an Action.Name.  Is there an easy way to get the translated versions of those or do you have to just use an ActionLink?  (FWIW, old BuffWindow used just the untranslated name in its output.  I don't know if there was a reason it didn't use ActionLinks.)

File: src/parser/jobs/brd/modules/RagingStrikes.tsx [hunk 17]
```diff
@@ -34,58 +43,97 @@ interface MuseWindow {
...
-	protected override init() {
-		super.init()
-
-		this.SUPPORT_ACTIONS = SUPPORT_ACTIONS.map(actionKey => this.data.actions[actionKey].id)
-		this.addEventHook('applybuff', {to: 'player', abilityId: [this.data.statuses.ARMYS_MUSE.id]}, this.onApplyMuse)
-		this.addEventHook('removebuff', {to: 'player', abilityId: [this.data.statuses.ARMYS_MUSE.id]}, this.onRemoveMuse)
+			adjustCount: this.adjustExpectedActionCount.bind(this),
+			adjustOutcome: this.adjustExpectedActionOutcome.bind(this),
+			barrageId: this.data.actions.BARRAGE.id,
+			wasBarrageUsed: this.wasBarageUsed.bind(this),
```
Comment (hintxiv) [inline src/parser/jobs/brd/modules/RagingStrikes.tsx:135]:
```suggestion
			wasBarrageUsed: this.wasBarrageUsed.bind(this),
```

File: src/parser/jobs/brd/modules/RagingStrikes.tsx [hunk 18]
```diff
@@ -170,4 +211,15 @@ export default class RagingStrikes extends BuffWindowModule {
 			}
 		}
 	}
+
+	private wasBarageUsed(window: HistoryEntry<EvaluatedAction[]>) {
```
Comment (hintxiv) [inline src/parser/jobs/brd/modules/RagingStrikes.tsx:215]:
```suggestion
	private wasBarrageUsed(window: HistoryEntry<EvaluatedAction[]>) {
```

**Review by hintxiv — CHANGES_REQUESTED** [review]
Really cool stuff. I didn't dive too deep into all of it but I can confirm that the RS window changes duplicate the old functionality on my test logs.

File: src/parser/jobs/nin/modules/TrickAttackWindow.tsx [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+
+	override generateNotes(window: HistoryEntry<EvaluatedAction[]>) {
+		return window.data.find(cast => cast.action.id === this.tcjId) ?
+			<Trans id="nin.taWindow.chart.notes.yes">Yes</Trans> :
+			<Trans id="nin.taWindow.chart.notes.no">No</Trans>
+	}
+}
+
+export class TrickAttackWindow extends BuffWindow {
+	static override handle = 'taWindow2'
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/jobs/nin/modules/TrickAttackWindow.tsx:56]: Yeah, leftover from when I had both a new and old file in the working copy at one point.

File: src/parser/jobs/drg/modules/DragonSight.tsx [hunk 20]
```diff
@@ -166,7 +198,7 @@ export default class DragonSight extends BuffWindowModule {
 		// mark it
 		if (actualWindowDuration < (windowDurationMillis - SHORT_WINDOW_BUFFER)) {
 			// activate the header because something's up
-			this.rotationTableNotesColumnHeader = <Trans id="drg.ds.notes.header">Short Window Cause</Trans>
+			//this.rotationTableNotesColumnHeader = <Trans id="drg.ds.notes.header">Short Window Cause</Trans>
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/jobs/drg/modules/DragonSight.tsx:201]: Yeah, another commented out while rearranging and forgot to delete.

File: src/parser/jobs/drg/modules/DragonSight.tsx (hunk 3 above)
Comment (KelosPelmand) [reply src/parser/jobs/drg/modules/DragonSight.tsx:87]: Surprisingly, none of the DRG parses I tested pointed this out, and I was testing from new parses, not just top parses or something.

File: src/parser/core/modules/ActionWindow/evaluators/AllowedGcdsOnlyEvaluator.tsx [hunk 21]
```diff
@@ -0,0 +1,105 @@
+import {Plural, Trans} from '@lingui/react'
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/evaluators/AllowedGcdsOnlyEvaluator.tsx:1]: Are you thinking one index file in ActionWindow folder or one each in ActionWindow/windows and ActionWindow/evaluators?

File: src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+		this.defaultEventHook = this.addEventHook(
+			filter<Event>()
+				.source(this.parser.actor.id)
+				.action(oneOf(actionsToTrack))
+				.type('action'),
+			this.onWindowAction)
+	}
+
+	override initialise() {
+		this.defaultEventHook = this.addEventHook(filter<Event>().source(this.parser.actor.id).type('action'), this.onWindowAction)
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx:131]: Something definitely would have to change on the interface, since derived classes can set their own hooks.  I have an idea or two about how to do it, so I'll see what makes sense.

File: src/parser/core/modules/ActionWindow/readme.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ackwell) [reply src/parser/core/modules/ActionWindow/readme.md:1]: Makes sense.

---

Haha, I'm just being snarky. It's style preference.

File: src/parser/core/modules/History.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (ackwell) [reply src/parser/core/modules/History.ts:95]: Yeah that's 👌. Bool chain expressions are a bit of a js-ism, i suppose.

File: src/parser/core/modules/History.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (ackwell) [reply src/parser/core/modules/History.ts:18]: I'm trying to get utils out of that folder, too 😬.

I think a sister folder to `modules` will be the final resting place for this stuff, but It's not quite _there_ yet - and I'd rather keep the folders clean in the interim. If it's needed soon, it's something we can look into after this or something.

File: src/parser/core/modules/ActionWindow/evaluators/AllowedGcdsOnlyEvaluator.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (ackwell) [reply src/parser/core/modules/ActionWindow/evaluators/AllowedGcdsOnlyEvaluator.tsx:1]: If we were going full-on namespace control, I'd go for `ActionWindow/index`, `ActionWindow/windows/index` and `ActionWindow/evaluators/index` - but realistically I don't think that's nessecary in this instance. `ActionWindow/index` to just define the public-facing API should be enough.

The Timeline module has an example of this, actually. forgot to post in original comment.

File: src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (ackwell) [reply src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx:131]: Probably the easiest way is to replace the default hook property with something that accepts an Event filter object. That way you can add/remove the hook and pass the current filter object in unconditionally.

Now that I think about it more, that also simplifies the overriding story - you no longer have to remember to call a remove before adding your own, can just override the filter property and it inherently replaces the default.

File: src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (ackwell) [reply src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx:142]: It's _really_ touchy-feely - I wish i had a hard rule that i could pass on but i really don't.

If anything, it's probably something along the line of a max of 1 on hot loops and 1-2 on cold ones - but this is really a CTA for me to look into some lazy chain tooling, as well.

File: src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (ackwell) [reply src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx:191]: Yeah, classic case of JS's really bad syntax decisions causing ambiguous usage of symbols. Without the paren, it's parsed as a code block.

File: src/parser/core/modules/ActionWindow/evaluators/LimitedActionsEvaluator.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (ackwell) [reply src/parser/core/modules/ActionWindow/evaluators/LimitedActionsEvaluator.tsx:47]: Hm. Not sure on historical context there. The names... probably should be removed honestly (from the data, that is) - there's no way to get them as translated values.

I've not tested (work 😭) if they work in-place (should do), but data links are the "correct" way of getting translated game data such as action names. you can disable the icon to keep it as just the name for ux purposes.

File: src/parser/core/modules/ActionWindow/windows/BuffWindow.ts [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+	abstract buffStatus: Status | Status[]
+
+	/**
+	 * Determines if a window ended early due to the end of the pull.
+	 * @param window The window to check
+	 * @returns True if the window is shorter than the expected duration of the buff because of the end
+	 * of the pull; false otherwise.
+	 */
+	protected isRushedEndOfPullWindow(window: HistoryEntry<EvaluatedAction[]>) {
+		const expectedDuration = (ensureArray(this.buffStatus)[0].duration ?? 0)
```
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/windows/BuffWindow.ts:34]: I copied this from old BuffWindow, but nobody seems to actually be using this potential feature at this time, so I think I'll remove the array option for now, and we can make sure we implement it right when someone does need it.

File: src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx:131]: I get the idea here, but I'm not sure what type I need to declare for the variable that will hold the filter.  Based on the `addEventHook` method, my guess would be `Event['type'] | Partial<Event> | EventFilterPredicate<Event>`

File: src/parser/core/modules/ActionWindow/evaluators/LimitedActionsEvaluator.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (KelosPelmand) [reply src/parser/core/modules/ActionWindow/evaluators/LimitedActionsEvaluator.tsx:47]: image

ActionLink with showIcon={false} does appear to work here.

**Comment by KelosPelmand** [comment]
Requested changes are in.

File: src/parser/core/modules/ActionWindow/evaluators/NotesEvaluator.ts [hunk 24]
```diff
@@ -9,9 +9,7 @@ export abstract class NotesEvaluator implements WindowEvaluator {
 	protected abstract header : RotationTarget
 	protected abstract generateNotes(window: HistoryEntry<EvaluatedAction[]>): JSX.Element
 
-	// Derived classes may decide to suggest, in which case they need the windows.
-	// eslint-disable-next-line @typescript-eslint/no-unused-vars
-	suggest(windows: Array<HistoryEntry<EvaluatedAction[]>>): Suggestion | undefined { return undefined }
+	suggest(_windows: Array<HistoryEntry<EvaluatedAction[]>>): Suggestion | undefined { return undefined }
```
Comment (ackwell) [inline src/parser/core/modules/ActionWindow/evaluators/NotesEvaluator.ts:12]: oh, well remembered. i'd forgotten that rule

File: src/parser/core/modules/ActionWindow/evaluators/AllowedGcdsOnlyEvaluator.tsx [hunk 25]
```diff
@@ -1,9 +1,9 @@
 import {Plural, Trans} from '@lingui/react'
 import React from 'react'
 import {GlobalCooldown} from '../../GlobalCooldown'
-import {HistoryEntry} from '../../History'
 import {SeverityTiers, TieredSuggestion} from '../../Suggestions/Suggestion'
 import {EvaluatedAction} from '../EvaluatedAction'
+import {HistoryEntry} from '../History'
```
Comment (ackwell) [inline src/parser/core/modules/ActionWindow/evaluators/AllowedGcdsOnlyEvaluator.tsx:6]: thanks - i'll try to get the folder worked out soon

**Review by ackwell — APPROVED** [review]
whew, thanks for getting all that in!

**Comment by ackwell** [comment]
Just realised it's #1300, nice
